patterns/observers/observer.cs

```diff
@@ -0,0 +1,102 @@
+using System;
+using System.Collections.Generic;
+
+// Subject interface
+public interface ISubject
+{
+    void Attach(IObserver observer);
+    void Detach(IObserver observer);
+    void Notify();
+}
+
+// Observer interface
+public interface IObserver
+{
+    void Update(int number);
+}
+// Concrete subject
+public class NumberAggregator : ISubject
+{
+    private List<int> numbers = new List<int>();
+    private List<IObserver> observers = new List<IObserver>();
+
+    public void Attach(IObserver observer)
+    {
+        observers.Add(observer);
+    }
+
+    public void Detach(IObserver observer)
+    {
+        observers.Remove(observer);
+    }
+
+    public void Notify()
+    {
+        foreach (var observer in observers)
+        {
+            observer.Update(numbers[numbers.Count - 1]);
+        }
+    }
+
+    public void AddNumber(int number)
+    {
+        numbers.Add(number);
+        Notify(); // Notify observers about the change
+    }
+
+    public double CalculateAverage()
+    {
+        if (numbers.Count == 0)
+            return 0;
+
+        int sum = 0;
+        foreach (var num in numbers)
+        {
+            sum += num;
+        }
+
+        return (double)sum / numbers.Count;
+    }
+}
+
+// Concrete observer
+public class AverageDisplay : IObserver
+{
+    private NumberAggregator aggregator;
```

Comment:
You've created an unnecessary dependency here that should not be present in an Observer pattern implementation. The observers themselves should not have direct reference to the subject. The subject holds a list of all observers so that it can update them without need of individual observers have direct reference to the subject.

The code is quite highly coupled: Since your observers have a reference to the subject, the observer's `Update` method actually calls on the subject's `CalculateAverage` method. This should be avoided since the result is:
| Subject --> | Subject --> | Observer --> | Subject |
| --- | --- | --- | --- |
| calls `AddNumber` | calls `Notify` | calls `Update` | calls `CalculateAverage` |

we shouldn't see this type of back and forth across subject and observer: On a particular event, or state change, subject should call notify and all observers should receive the update, but should **not** be calling back to subject in turn.